test/model/message_test.dart

```diff
@@ -123,6 +127,269 @@ void main() {
+      // First request has succeeded; event hasn't arrived
+      // Mid-second request
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isFalse();
+      check(store.getEditMessageErrorStatus(otherMessage.id)).isNotNull().isFalse();
+
+      // First event arrives
+      await store.handleEvent(eg.updateMessageEditEvent(message));
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+      checkNotifiedOnce();
+
+      async.elapse(Duration(milliseconds: 500));
+      // Second request has succeeded; event hasn't arrived
+      check(store.getEditMessageErrorStatus(otherMessage.id)).isNotNull().isFalse();
+
+      // Second event arrives
+      await store.handleEvent(eg.updateMessageEditEvent(otherMessage));
+      check(store.getEditMessageErrorStatus(otherMessage.id)).isNull();
+      checkNotifiedOnce();
+    }));
+
+    test('request fails', () => awaitFakeAsync((async) async {
+      await prepareEditMessage();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(apiException: eg.apiBadRequest(), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      checkNotifiedOnce();
+      async.elapse(Duration(seconds: 1));
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isTrue();
+      checkNotifiedOnce();
+    }));
+
+    test('request fails; take failed edit', () => awaitFakeAsync((async) async {
+      await prepareEditMessage();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(apiException: eg.apiBadRequest(), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      checkNotifiedOnce();
+      async.elapse(Duration(seconds: 1));
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isTrue();
+      checkNotifiedOnce();
+
+      check(store.takeFailedMessageEdit(message.id)).equals('new content');
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+      checkNotifiedOnce();
+    }));
+
+    test('takeFailedMessageEdit throws StateError when nothing to take', () => awaitFakeAsync((async) async {
+      await prepareEditMessage();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+      check(() => store.takeFailedMessageEdit(message.id)).throws<StateError>();
+    }));
+
+    test('editMessage throws StateError if editMessage already in progress for same message', () => awaitFakeAsync((async) async {
+      await prepareEditMessage();
+
+      connection.prepare(
+        json: UpdateMessageResult().toJson(), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      async.elapse(Duration(milliseconds: 500));
+      check(connection.takeRequests()).length.equals(1);
+      checkNotifiedOnce();
+
+      connection.prepare(json: UpdateMessageResult().toJson());
```

Comment:
I'm a bit surprised that this is needed, since I thought `store.editMessage` would fail before firing the request, which is confirmed by the later `.isEmpty` check on `takeRequests()`. If I missed something, maybe a comment here will be helpful?

Reply:
Ah yeah, this can be removed.